File: lib/player.rb
#!/usr/bin/env ruby
# encoding: utf-8

class Spotbot::Player
  attr_reader :queue, :playlist, :plaything, :support, :logger, :current_track, :changing_track

  def initialize(logger)
    @queue = Spotbot::Queue.instance
    @playlist = Spotbot::Playlist.instance
    @logger = logger
    @plaything = Plaything.new
    @changing_track = false
  end

  def run
    @support = Spotbot::SpotifySupport.instance
    Spotbot::SpotifySupport::DEFAULT_CONFIG[:callbacks] = Spotify::SessionCallbacks.new(session_callbacks)

    support.initialize_spotify!

    play_next

    EM.defer do
      support.poll { } # keep the player running
    end
  end

  def play(uri = nil)
    play_track(uri) if uri
    play_next unless current_track
    Spotify.try(:session_player_play, support.session, true)
  end

  def pause
    Spotify.try(:session_player_play, support.session, false)
  end

  def play_next
    if track = next_track
      play_track(track)
      track
    end
  end

  private

  def next_track
    queue.next || playlist.next
  end

  def play_track(uri)
    @changing_track = true
    logger.info("play_track") { uri }

    link = Spotify.link_create_from_string(uri)
    track = Spotify.link_as_track(link)
    support.poll { Spotify.track_is_loaded(track) }
    self.current_track = uri

    # Pause before we load a new track. Fixes a quirk in libspotify.
    Spotify.try(:session_player_play, support.session, false)
    Spotify.try(:session_player_load, support.session, track)
    Spotify.try(:session_player_play, support.session, true)
    @changing_track = false
  end

  def current_track=(uri)
    @current_track = uri
    Spotbot::Firebase.current_track(uri) if uri
  end

  def session_callbacks
    @session_callbacks ||= {
      log_message: proc do |session, message|
        logger.info("session (log message)") { message }
      end,

      logged_in: proc do |session, error|
        logger.debug("session (logged in)") { Spotify::Error.explain(error) }
      end,

      logged_out: proc do |session|
        logger.debug("session (logged out)") { "logged out!" }
      end,

      streaming_error: proc do |session, error|
        logger.error("session (player)") { "streaming error %s" % Spotify::Error.explain(error) }
      end,

      start_playback: proc do |session|
        logger.debug("session (player)") { "start playback" }
        plaything.play
      end,

      stop_playback: proc do |session|
        logger.debug("session (player)") { "stop playback" }
        plaything.stop
      end,

      get_audio_buffer_stats: proc do |session, stats|
        stats[:samples] = plaything.queue_size
        stats[:stutter] = plaything.drops
        logger.debug("session (player)") { "queue size [#{stats[:samples]}, #{stats[:stutter]}]" }
      end,

      music_delivery: proc do |session, format, frames, num_frames|
        if num_frames == 0
          logger.debug("session (player)") { "music delivery audio discontuity" }
          plaything.stop
          0
        else
          frames = FrameReader.new(format[:channels], format[:sample_type], num_frames, frames)
          consumed_frames = plaything.stream(frames, format.to_h)
          logger.debug("session (player)") { "music delivery #{consumed_frames} of #{num_frames}" }
          consumed_frames
        end
      end,

      end_of_track: proc do |session|
        unless changing_track
          logger.debug("session (player)") { "end of track" }
          self.current_track = nil
          play_next
        end
      end,
    }
  end
end

File: lib/player.js
"use strict";
module.exports = function (spotify, queue, playlist) {
  var firebase = require("./firebase");
  var loudness = require("loudness");
  var lastfm = require("./lastfm");
  var firebaseStatus = firebase.ref.child("player/playing");
  var firebaseNext = firebase.ref.child("player/next");
  var firebaseCurrent = firebase.ref.child("player/current_track");
  var firebaseVolume = firebase.ref.child("volume");
  var Track = require("./track")(spotify);
  var currentTrack = null;

  firebaseStatus.on("value", function(data){
    console.log("Status updated", data.val());
    if (data.val() === true) {
      console.log("Player play");
      firebaseStatus.set(true);
      spotify.player.resume();
    } else if (data.val() === false) {
      pause();
    }
  });

  firebaseNext.on("value", function(data){
    console.log("Next updated", data.val());
    if (data.val() === true) {
      playNext();
      firebaseNext.set(false);
    }
  });

  firebaseCurrent.on("child_changed", function(data){
    var track = new Track(data.val());
    if(typeof(track.uri) !== "undefined") {
      if (!currentTrack || track.uri !== currentTrack.uri ) {
        console.log("Change track", track.uri);
        playNext(track);
      }
    }
  });

  firebaseVolume.on("value", function(data){
    if (data.val() !== null) {
      console.log("Set Volume to", data.val());
      loudness.setVolume(data.val(), function (err) {
        if (err) {
          console.log("Error", err);
        }
      });
    }
  });

  function start(){
    console.log("Starting player");
    firebaseStatus.set(true);

    setCurrentTrack();

    spotify.player.on({
        endOfTrack: scrobbleAndPlayNext
    });

    playNext();
  }

  function pause(){
    console.log("Player pause");
    firebaseStatus.set(false);
    spotify.player.pause();
  }

  function stop(){
    pause();
    setCurrentTrack();
  }

  function scrobbleAndPlayNext(){
    lastfm.scrobble(currentTrack);
    playNext();
  }

  function playNext(track){
    var trackToPlay = track || nextTrack();
    if (trackToPlay) {
      console.log("Playing track:", trackToPlay.uri);
      try {
        spotify.player.play(trackToPlay.getLink());
        firebaseStatus.set(true);
        setCurrentTrack(trackToPlay);
      } catch (e) {
        console.log(e);
        playNext();
      }
    } else {
      console.log("No track to play. Trying again in 1 second");
      pause();
      setTimeout(playNext, 1000);
    }
  }

  function nextTrack(){
    return queue.next() || playlist.next();
  }

  function setCurrentTrack(track){
    if (track) {
      currentTrack = track;
      firebaseCurrent.set({uri: track.uri, started_at: firebase.firebase.ServerValue.TIMESTAMP});

      if (track.isLoaded) {
        console.log("isLoaded");
        lastfm.nowPlaying(track);
      } else {
        setTimeout(function(){
          lastfm.nowPlaying(track);
        }, 1000);
      }
    } else {
      currentTrack = null;
      firebaseCurrent.remove();
    }
  }

  return {
    start: start,
    stop: stop,
  };
};
